R/check-nonmem-finished.R

```diff
@@ -0,0 +1,30 @@
+#' Check if a NONMEM Bayes model has finished running
+#'
+#' This function implements the [bbr::check_nonmem_finished()] method for NONMEM
+#' Bayes models. To decide if a model is finished, it runs
+#' [bbr::check_nonmem_finished()] on each of the chain submodels.
+#'
+#' Note that, like the `bbi_nonmem_model` method, this method is intended to be
+#' called on models that have been submitted. It happens to also return `TRUE`
+#' for un-submitted models, but that should be considered undefined behavior.
+#'
+#' @param .mod A `bbi_nmbayes_model` object.
+#' @param ... Arguments passed to method call on each chain submodel.
+#' @return `TRUE` if all of the chain models have finished, and `FALSE`
+#'   otherwise.
+#' @name nmbayes_check_nonmem_finished
+#' @export
+check_nonmem_finished.bbi_nmbayes_model <- function(.mod, ...) {
+  outdir <- get_output_dir(.mod, .check_exists = FALSE)
+  if (!fs::dir_exists(outdir)) {
+    # check_nonmem_finished.bbi_nonmem_model() takes this as an indication that
+    # submission failed right away. For nmbayes model the top-level output
+    # directory is created synchronously even when `.wait = FALSE`. So it is
+    # likely a model without an output directory just wasn't submitted, but
+    # return `TRUE` for consistency.
```

Comment:
Side remark 1: I was surprised by `check_nonmem_finished` returning `TRUE` for un-submitted NONMEM models, but (without looking into it too deeply) I guess the problem is that it can't distinguish between un-submitted and "failed right away when `.wait = FALSE`".

Side remark 2: for regular NONMEM models, I would guess that the "output directory exists?" guard introduces a race condition between `check_nonmem_finished` and `bbi` creating the output directory.  Even if that's true, perhaps there aren't any easy/good alternatives here and it's good enough.

Reply:
> I was surprised by check_nonmem_finished returning TRUE for un-submitted NONMEM models, but (without looking into it too deeply) I guess the problem is that it can't distinguish between un-submitted and "failed right away when `.wait = FALSE`".

Yep, that's right. We could consider re-implementing this with a different method of checking, but that's out of scope for these changes. (And would technically be a breaking change.) A more accurate name might've been `check_nonmem_not_running()`, but that's kind of a mouthful and probably confusing to most users.

> I would guess that the "output directory exists?" guard introduces a race condition... Even if that's true, perhaps there aren't any easy/good alternatives here and it's good enough.

That's true, though I don't think we've had issues with this in practice. I can't currently think of any alternatives, but I'm certainly open to them.

Reply:
Regardless of all that, I appreciate the code comment [here](https://github.com/metrumresearchgroup/bbr.bayes/blob/69b507efa052f63bdcea10b64720e4bc7bd04b1a/R/check-nonmem-finished.R#L20-L24).

Reply:
> Yep, that's right. We could consider re-implementing this with a different method of checking, but that's out of scope for these changes. 

Right, to be clear, these are all just side remarks.  I'm not proposing we change anything in relation to them.  They can even be ignored entirely, though knowing what suprised may be helpful for why I worded something the docs/comments the way I did.

I do think it would be very helpful for the check_nonmem_finished docs to explain that (something along the lines of what I put in this methods docstring).

> (And would technically be a breaking change)

Perhaps.  It sounds like we're into undefined behavior territory here, and I don't think it be out of the question to error or return NA here, but again, not really relevant for this PR.